### PR TITLE
Prioritise based on attestation subnets when connecting and disconnecting peers

### PR DESCRIPTION
**Pull request by ajsutton**
## PR Description
To be able to publish attestations, Teku needs to ensure it has at least one peer connected that is subscribed to each attestation subnet.  To help do that when connecting to or disconnecting from peer, we rank the peers based on the subnets they are subscribed to and how many other peers we have already subscribed to that subnet.

When connecting, we only have the persistent subnets advertised in the peer's ENR to base the decision off but once connected we use the peer's actual subscriptions, which includes persistent and short term subscriptions. This ensures that the subscriptions we score a peer on are always up to date even if we don't have their latest ENR and avoids issues with peers advertising a lot of subnets in their ENR but not actually subscribing to them.

The scoring system is designed to give high scores to peers that subscribe to subnets with no or few subscribers and very quickly reduce the score boost for subnets that have more peers. It's a pretty arbitrary algorithm but seems to give reasonable results and is definitely better than randomly picking peers.

To fully ensure we have all subnets covered, we should also continue attempting to connect to peers even if we have reached our peer limit if there are subnets with limited subscribers and discovered peers claim to subscribe to those subnets.  Potentially we could also retry sending attestations for a short period if we don't have a peer to send them to when first produced. Those changes are for follow up PRs.

## Fixed Issue(s)
Part of #1963

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

### CODE REVIEW

File: networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/connection/ConnectionManager.java [hunk 1]
```diff
@@ -103,9 +109,16 @@ public ConnectionManager(
...
+    if (maxAttempts == 0) {
+      return;
+    }
+    final PeerScorer peerScorer = peerScorerFactory.create();
     discoveryService
         .streamKnownPeers()
         .filter(this::isPeerValid)
+        .sorted(
+            Comparator.comparing((Function<DiscoveryPeer, Integer>) peerScorer::scoreCandidatePeer)
+                .reversed())
```
Comment (Nashatyrev) [inline networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/connection/ConnectionManager.java:121]: May be it makes sense to add some randomness here to make connection manager more reliable? I.e. take 50% of peers by score and 50% of random peers.
It occurred several times from my former experience that strictly selecting peers by their score may lead to connectivity problems due to different reasons

File: networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/connection/ConnectionManager.java [hunk 2]
```diff
@@ -148,9 +161,14 @@ private void attemptConnection(final PeerAddress discoveryPeer) {
...
   private void onPeerConnected(final Peer peer) {
     final int peersToDrop = targetPeerCountRange.getPeersToDrop(network.getPeerCount());
+    if (peersToDrop == 0) {
+      return;
+    }
+    final PeerScorer peerScorer = peerScorerFactory.create();
     network
         .streamPeers()
         .filter(candidate -> !staticPeers.contains(candidate.getAddress()))
+        .sorted(Comparator.comparing(peerScorer::scoreExistingPeer))
```
Comment (Nashatyrev) [inline networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/connection/ConnectionManager.java:171]: The same note on randomness here

File: networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/gossip/LibP2PGossipNetwork.java [hunk 3]
```diff
@@ -50,4 +60,24 @@ public TopicChannel subscribe(final String topic, final TopicHandler topicHandle
     PubsubSubscription subscription = gossip.subscribe(gossipHandler, libP2PTopic);
     return new LibP2PTopicChannel(gossipHandler, subscription);
   }
+
+  @Override
+  public Map<String, Collection<NodeId>> getSubscribersByTopic() {
+    return gossip
+        .getRouter()
```
Comment (Nashatyrev) [inline networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/gossip/LibP2PGossipNetwork.java:67]: May be consider including this feature to public Pubsub API?

**Review by Nashatyrev — APPROVED** [review]
Looks reasonable 👍 

BTW you may get additional info from this Gossip callback: https://github.com/libp2p/jvm-libp2p/blob/3d5be408f6cc4195858cdabef187160bd812a3d8/src/main/kotlin/io/libp2p/pubsub/gossip/GossipParams.kt#L180-L184
This is a new `Peer Exchange` feature of Gossip 1.1
However I'm not sure if it would be too useful cause we already have discovery ENR records

File: networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/gossip/LibP2PGossipNetwork.java [hunk 4]
```diff
@@ -50,4 +60,24 @@ public TopicChannel subscribe(final String topic, final TopicHandler topicHandle
...
+        .submitOnEventThread(
```
Comment (Nashatyrev) [inline networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/gossip/LibP2PGossipNetwork.java:68]: 👍

File: networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/connection/ConnectionManager.java (hunk 1 above)
Comment (ajsutton) [reply networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/connection/ConnectionManager.java:121]: Seems like a good idea, logged https://github.com/PegaSysEng/teku/issues/2332 to follow up and do that.

File: networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/gossip/LibP2PGossipNetwork.java (hunk 3 above)
Comment (ajsutton) [reply networking/p2p/src/main/java/tech/pegasys/teku/networking/p2p/libp2p/gossip/LibP2PGossipNetwork.java:67]: Yeah seems like a good idea.  Will merge it here now, raise a PR to jvm-libp2p and will then use the libp2p version once we pull in its next release.

**Comment by ajsutton** [comment]
Logged https://github.com/PegaSysEng/teku/issues/2331 to investigate supporting LibP2P peer exchange.

**Pull request merged**